File: Client/src/main/java/com/rspsi/editor/input/EditorInputRouter.java
package com.rspsi.editor.input;

import com.rspsi.editor.plugin.EditorPluginContext;
import com.rspsi.editor.tool.EditorToolController;

import java.util.Objects;

/**
 * Routes frontend-translated input through the neutral editor contracts.
 *
 * <p>JavaFX, Dear ImGui, and future frontends may disagree about how focus is
 * represented, but they must agree on this ordering: a focused text editor
 * keeps text-editing keys, otherwise registered editor shortcuts get the first
 * chance to handle a pressed key. Pointer events continue to flow to the
 * active tool controller.</p>
 */
public final class EditorInputRouter implements PointerEventSink {
    private final EditorPluginContext pluginContext;
    private final EditorToolController tools;

    public EditorInputRouter(EditorPluginContext pluginContext,
                             EditorToolController tools) {
        this.pluginContext = Objects.requireNonNull(pluginContext, "pluginContext");
        this.tools = Objects.requireNonNull(tools, "tools");
    }

    /** Dispatches a key when no frontend text editor currently owns focus. */
    public boolean key(EditorKeyEvent event) {
        return key(event, false);
    }

    /**
     * Dispatches one translated key event.
     *
     * @param textInputFocused true when a text field owns keyboard input;
     *                         global editor shortcuts are then skipped
     */
    public boolean key(EditorKeyEvent event, boolean textInputFocused) {
        Objects.requireNonNull(event, "event");
        if (textInputFocused || !event.pressed()) return false;
        return pluginContext.registry().dispatchShortcut(pluginContext, event);
    }

    public EditorToolController tools() {
        return tools;
    }

    @Override
    public void pointerDown(PointerEvent event) {
        tools.pointerDown(event);
    }

    @Override
    public void pointerDrag(PointerEvent event) {
        tools.pointerDrag(event);
    }

    @Override
    public void pointerUp(PointerEvent event) {
        tools.pointerUp(event);
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/input/EditorInputRouter.kt
package com.rspsi.editor.input

import com.rspsi.editor.plugin.EditorPluginContext
import com.rspsi.editor.tool.EditorToolController

/**
 * Routes frontend-translated input through neutral editor contracts.
 *
 * Frontends may disagree about focus representation, but the neutral ordering is fixed:
 * a focused text editor keeps keyboard input; otherwise pressed keys may trigger registered
 * editor shortcuts. Pointer lifecycle events are forwarded unchanged to the active tool.
 *
 * This router deliberately does not interpret gestures or mutate editor state. That keeps
 * frontend translation separate from future sealed tool-event/state-machine work.
 */
class EditorInputRouter(
    pluginContext: EditorPluginContext?,
    tools: EditorToolController?,
) : PointerEventSink {
    private val pluginContextValue =
        pluginContext ?: throw NullPointerException("pluginContext")
    private val toolsValue =
        tools ?: throw NullPointerException("tools")

    /** Dispatches a key when no frontend text editor currently owns focus. */
    fun key(event: EditorKeyEvent?): Boolean = key(event, false)

    /**
     * Dispatches one translated key event.
     *
     * When [textInputFocused] is true, global editor shortcuts are intentionally skipped
     * so normal text editing remains owned by the frontend control with focus.
     */
    fun key(
        event: EditorKeyEvent?,
        textInputFocused: Boolean,
    ): Boolean {
        val safeEvent = event ?: throw NullPointerException("event")
        if (textInputFocused || !safeEvent.pressed()) {
            return false
        }
        return pluginContextValue.registry().dispatchShortcut(pluginContextValue, safeEvent)
    }

    fun tools(): EditorToolController = toolsValue

    override fun pointerDown(event: PointerEvent?) {
        toolsValue.pointerDown(event)
    }

    override fun pointerDrag(event: PointerEvent?) {
        toolsValue.pointerDrag(event)
    }

    override fun pointerUp(event: PointerEvent?) {
        toolsValue.pointerUp(event)
    }

    override fun pointerMove(event: PointerEvent?) {
        toolsValue.pointerMove(event)
    }
}
